// @jupyterlab/services/lib/setting/index.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
Object.defineProperty(exports, "__esModule", { value: true });
var coreutils_1 = require("@jupyterlab/coreutils");
var serverconnection_1 = require("../serverconnection");
/**
 * The url for the lab settings service.
 */
var SERVICE_SETTINGS_URL = 'lab/api/settings';
/**
 * The static namespace for `SettingManager`.
 */
var SettingManager = (function () {
    /**
     * Create a new setting manager.
     */
    function SettingManager(options) {
        if (options === void 0) { options = {}; }
        this.serverSettings = options.serverSettings ||
            serverconnection_1.ServerConnection.makeSettings();
    }
    /**
     * Fetch a plugin's settings.
     *
     * @param id - The plugin's ID.
     *
     * @returns A promise that resolves with the plugin settings or rejects
     * with a `ServerConnection.IError`.
     */
    SettingManager.prototype.fetch = function (id) {
        var base = this.serverSettings.baseUrl;
        var request = { method: 'GET', url: Private.url(base, id) };
        var serverSettings = this.serverSettings;
        var promise = serverconnection_1.ServerConnection.makeRequest(request, serverSettings);
        return promise.then(function (response) {
            var status = response.xhr.status;
            if (status !== 200) {
                throw serverconnection_1.ServerConnection.makeError(response);
            }
            return response.data;
        }).catch(function (reason) { throw serverconnection_1.ServerConnection.makeError(reason); });
    };
    /**
     * Save a plugin's settings.
     *
     * @param id - The plugin's ID.
     *
     * @param user - The plugin's user setting values.
     *
     * @returns A promise that resolves when saving is complete or rejects
     * with a `ServerConnection.IError`.
     */
    SettingManager.prototype.save = function (id, user) {
        var base = this.serverSettings.baseUrl;
        var request = {
            data: JSON.stringify(user),
            contentType: 'application/json',
            method: 'PATCH',
            url: Private.url(base, id)
        };
        var serverSettings = this.serverSettings;
        var promise = serverconnection_1.ServerConnection.makeRequest(request, serverSettings);
        return promise.then(function (response) {
            var status = response.xhr.status;
            if (status !== 204) {
                throw serverconnection_1.ServerConnection.makeError(response);
            }
            return void 0;
        }).catch(function (reason) { throw serverconnection_1.ServerConnection.makeError(reason); });
    };
    return SettingManager;
}());
exports.SettingManager = SettingManager;
/**
 * A namespace for private data.
 */
var Private;
(function (Private) {
    /**
     * Get the url for a plugin's settings.
     */
    function url(base, id) {
        return coreutils_1.URLExt.join(base, SERVICE_SETTINGS_URL, id);
    }
    Private.url = url;
})(Private || (Private = {}));
